File: benchmarks/Java/src/richards/TaskControlBlock.java
/*
 * This benchmark is derived from Mario Wolczko's Java and Smalltalk version of
 * Richards.
 * 
 * It is modified based on the SOM version and to use Java 8 features.
 * License details:
 *   http://web.archive.org/web/20050825101121/http://www.sunlabs.com/people/mario/java_benchmarking/index.html
 */
package richards;


final class TaskControlBlock extends TaskState {
  private final TaskControlBlock link;
  private final int identity;
  private final int priority;
  private Packet input;
  private final ProcessFunction function;
  private final RBObject handle;

  TaskControlBlock(final TaskControlBlock aLink, final int anIdentity,
      final int aPriority, final Packet anInitialWorkQueue,
      final TaskState anInitialState, final ProcessFunction aBlock,
      final RBObject aPrivateData) {
    link = aLink;
    identity = anIdentity;
    priority = aPriority;
    input = anInitialWorkQueue;
    setPacketPending(anInitialState.isPacketPending());
    setTaskWaiting(anInitialState.isTaskWaiting());
    setTaskHolding(anInitialState.isTaskHolding());
    function = aBlock;
    handle = aPrivateData;
  }

  public int getIdentity() { return identity; }
  public TaskControlBlock getLink()  { return link; }
  public int getPriority() { return priority; }

  public TaskControlBlock addInputAndCheckPriority(final Packet packet,
      final TaskControlBlock oldTask) {
    if (NO_WORK == input) {
      input = packet;
      setPacketPending(true);
      if (priority > oldTask.getPriority()) { return this; }
    } else {
      input = append(packet, input);
    }
    return oldTask;
  }

  public TaskControlBlock runTask() {
    Packet message;
    if (isWaitingWithPacket()) {
      message = input;
      input = message.getLink();
      if (NO_WORK == input) {
        running();
      } else {
        packetPending();
      }
    } else {
      message = NO_WORK;
    }
    return function.apply(message, handle);
  }
}

File: benchmarks/Rust/src/richards/task_control_block.rs
use crate::richards::packet::PacketBox;
use crate::richards::rb_object::RBObject;
use crate::richards::scheduler::Scheduler;
use crate::richards::task_state::TaskState;
use std::cell::{Cell, RefCell};
use std::rc::Rc;

pub struct TaskControlBlock {
    link: Option<TaskControlBlockRc>,
    identity: usize,
    priority: usize,
    input: Cell<Option<PacketBox>>,
    state: TaskState,
    #[allow(clippy::type_complexity)]
    function: Box<RefCell<dyn FnMut(&Scheduler, Option<PacketBox>) -> Option<TaskControlBlockRc>>>,
}

#[allow(clippy::module_name_repetitions)]
pub type TaskControlBlockRc = Rc<TaskControlBlock>;

impl RBObject for TaskControlBlock {}

impl TaskControlBlock {
    pub fn new(
        link: Option<TaskControlBlockRc>,
        identity: usize,
        priority: usize,
        initial_input: Option<PacketBox>,
        initial_state: TaskState,
        function: impl 'static + FnMut(&Scheduler, Option<PacketBox>) -> Option<TaskControlBlockRc>,
    ) -> Self {
        Self {
            link,
            identity,
            priority,
            state: initial_state,
            input: Cell::new(initial_input),
            function: Box::new(RefCell::new(function)),
        }
    }

    pub fn new_rc(
        link: Option<TaskControlBlockRc>,
        identity: usize,
        priority: usize,
        an_initial_work_queue: Option<PacketBox>,
        an_initial_state: TaskState,
        a_block: impl 'static + FnMut(&Scheduler, Option<PacketBox>) -> Option<TaskControlBlockRc>,
    ) -> TaskControlBlockRc {
        Rc::new(Self::new(
            link,
            identity,
            priority,
            an_initial_work_queue,
            an_initial_state,
            a_block,
        ))
    }

    pub fn get_identity(&self) -> usize {
        self.identity
    }

    pub fn get_link(&self) -> &Option<TaskControlBlockRc> {
        &self.link
    }

    pub fn get_priority(&self) -> usize {
        self.priority
    }

    pub fn set_task_holding(&self, hold: bool) {
        self.state.set_task_holding(hold);
    }

    pub fn set_task_waiting(&self, wait: bool) {
        self.state.set_task_waiting(wait);
    }

    pub fn is_task_holding_or_waiting(&self) -> bool {
        self.state.is_task_holding_or_waiting()
    }

    pub fn add_input_and_check_priority(
        &self,
        packet: PacketBox,
        old_task: &TaskControlBlockRc,
        this_task: &TaskControlBlockRc,
    ) -> TaskControlBlockRc {
        let input = self.input.replace(None);
        if input.is_none() {
            self.input.set(Some(packet));
            self.state.set_packet_pending(true);
            if self.priority > old_task.get_priority() {
                return this_task.clone();
            }
        } else {
            self.input.set(Some(Self::append(packet, input)));
        }
        old_task.clone()
    }

    pub fn run_task(&self, scheduler: &Scheduler) -> Option<TaskControlBlockRc> {
        let message = if self.state.is_waiting_with_packet() {
            let mut message = self.input.replace(None);
            match message.as_mut().unwrap().take_link() {
                None => {
                    self.input.set(None);
                    self.state.running();
                }
                Some(input) => {
                    self.input.set(Some(input));
                    self.state.packet_pending();
                }
            }
            message
        } else {
            Self::NO_WORK
        };
        self.function.borrow_mut()(scheduler, message)
    }
}
